parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/CompatibilityUtil.java

```diff
@@ -66,49 +82,73 @@ private V21FileAPI() throws ReflectiveOperationException {
     }
   }
 
-  private static Object invoke(Method method, String errorMsg, Object instance, Object... args) {
-    try {
-      return method.invoke(instance, args);
-    } catch (IllegalAccessException e) {
-      throw new IllegalArgumentException(errorMsg, e);
-    } catch (InvocationTargetException e) {
-      throw new IllegalArgumentException(errorMsg, e);
+  /**
+   * This method attempts to read into the provided readBuffer, readBuffer.remaining() bytes.
+   * If the underlying InputStream supports read directly into ByteBuffer we go ahead and invoke that.
+   * Else we fall back to directly calling readFully() on the underlying stream.
+   * @return Number of bytes read - should be readBuf.remaining()
+   */
+  public int getBuf(FSDataInputStream f, ByteBuffer readBuf) throws IOException {
+    int res;
+    if (useV2) {
+      res = readWithByteBuffer(f, readBuf);
+    } else {
+      if (readBuf.hasArray()) {
+        res = readWithExistingArray(f, readBuf);
+      } else {
+        res = readWithNewArray(f, readBuf);
+      }
     }
+    return res;
   }
 
-  public static int getBuf(FSDataInputStream f, ByteBuffer readBuf, int maxSize) throws IOException {
-    int res;
-    if (useV21) {
-      try {
-        res = (Integer) fileAPI.PROVIDE_BUF_READ_METHOD.invoke(f, readBuf);
-      } catch (InvocationTargetException e) {
-        if (e.getCause() instanceof UnsupportedOperationException) {
-          // the FSDataInputStream docs say specifically that implementations
-          // can choose to throw UnsupportedOperationException, so this should
-          // be a reasonable check to make to see if the interface is
-          // present but not implemented and we should be falling back
-          useV21 = false;
-          return getBuf(f, readBuf, maxSize);
-        } else if (e.getCause() instanceof IOException) {
-          throw (IOException) e.getCause();
-        } else {
-          // To handle any cases where a Runtime exception occurs and provide
-          // some additional context information. A stacktrace would just give
-          // a line number, this at least tells them we were using the version
-          // of the read method designed for using a ByteBuffer.
-          throw new IOException("Error reading out of an FSDataInputStream " +
-              "using the Hadoop 2 ByteBuffer based read method.", e.getCause());
-        }
-      } catch (IllegalAccessException e) {
-        // This method is public because it is defined in an interface,
-        // there should be no problems accessing it
-        throw new ShouldNeverHappenException(e);
+  private int readWithByteBuffer(FSDataInputStream f, ByteBuffer readBuf) throws IOException {
+    int remaining = readBuf.remaining();
+    try {
+      while (readBuf.hasRemaining()) {
```

Comment:
is there no readFully() method we can call that handles this for us? that seems surprising.